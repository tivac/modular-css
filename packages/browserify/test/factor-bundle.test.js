"use strict";

const fs     = require("fs");
const assert = require("assert");
const dedent     = require("dedent");
const browserify = require("browserify");
const from       = require("from2-string");
const shell      = require("shelljs");
const read   = require("@modular-css/test-utils/read.js")(__dirname);
const prefix = require("@modular-css/test-utils/prefix.js")(__dirname);
const bundle = require("./lib/bundle.js");
const plugin = require("../browserify.js");

describe("/browserify.js", () => {
    describe("factor-bundle", () => {
        beforeAll(() => shell.mkdir("./packages/browserify/test/output/factor-bundle"));
        afterAll(() => shell.rm("-rf", "./packages/browserify/test/output/factor-bundle"));
        
        it("should be supported", async () => {
            const build = browserify([
                    from(dedent(`
                        require('./packages/browserify/test/specimens/factor-bundle/basic/common.js');
                        require('./packages/browserify/test/specimens/start.css');
                    `)),
                    from(dedent(`
                        require('./packages/browserify/test/specimens/factor-bundle/basic/common.js');
                        require('./packages/browserify/test/specimens/local.css');
                    `)),
                ]);
            
            build.plugin(plugin, {
                css : prefix("output/factor-bundle/basic/basic.css"),
            });

            build.plugin(require("factor-bundle"), {
                outputs : [
                    prefix("output/factor-bundle/basic-a.js"),
                    prefix("output/factor-bundle/basic-b.js"),
                ],
            });
            
            const out = await bundle(build);
                
            expect(out).toMatchSnapshot();
            expect(read("./factor-bundle/basic/basic.css")).toMatchSnapshot();
            expect(read("./factor-bundle/basic/_stream_0.css")).toMatchSnapshot();
        });
        
        it("should support files w/o commonalities", async () => {
            const build = browserify([
                    from(dedent(`
                        require('./packages/browserify/test/specimens/simple.css');
                    `)),
                    from(dedent(`
                        require('./packages/browserify/test/specimens/blue.css');
                    `)),
                ]);
            
            build.plugin(plugin, {
                css : prefix("output/factor-bundle/nocommon/nocommon.css"),
            });

            build.plugin(require("factor-bundle"), {
                outputs : [
                    prefix("output/factor-bundle/nocommon-a.js"),
                    prefix("output/factor-bundle/nocommon-b.js"),
                ],
            });
            
            await bundle(build);

            expect(read("./factor-bundle/nocommon/_stream_0.css")).toMatchSnapshot();
            expect(read("./factor-bundle/nocommon/_stream_1.css")).toMatchSnapshot();
        });
        
        it("should properly handle files w/o dependencies", async () => {
            const build = browserify([
                    prefix("specimens/factor-bundle/deps/a.js"),
                    prefix("specimens/factor-bundle/deps/b.js"),
                ]);
            
            build.plugin(plugin, {
                css : prefix("output/factor-bundle/deps/deps.css"),
            });

            build.plugin(require("factor-bundle"), {
                outputs : [
                    prefix("output/factor-bundle/deps-a.js"),
                    prefix("output/factor-bundle/deps-b.js"),
                ],
            });
            
            await bundle(build);

            expect(read("./factor-bundle/deps/deps.css")).toMatchSnapshot();
            expect(read("./factor-bundle/deps/a.css")).toMatchSnapshot();
        });

        it("should support relative paths within factor-bundle files", async () => {
            const build = browserify([
                    prefix("specimens/factor-bundle/relative/a.js"),
                    prefix("specimens/factor-bundle/relative/b.js"),
                ]);
            
            build.plugin(plugin, {
                css : prefix("output/factor-bundle/relative/relative.css"),
            });

            build.plugin(require("factor-bundle"), {
                outputs : [
                    prefix("output/factor-bundle/relative-a.js"),
                    prefix("output/factor-bundle/relative-b.js"),
                ],
            });
            
            await bundle(build);

            expect(read("./factor-bundle/relative/relative.css")).toMatchSnapshot();
            expect(read("./factor-bundle/relative/a.css")).toMatchSnapshot();
        });

        it("should avoid outputting empty css files by default", async () => {
            const build = browserify([
                    prefix("specimens/factor-bundle/noempty/a.js"),
                    prefix("specimens/factor-bundle/noempty/b.js"),
                ]);
            
            build.plugin(plugin, {
                css : prefix("output/factor-bundle/noempty/noempty.css"),
            });

            build.plugin(require("factor-bundle"), {
                outputs : [
                    prefix("output/factor-bundle/noempty-a.js"),
                    prefix("output/factor-bundle/noempty-b.js"),
                ],
            });
            
            await bundle(build);

            assert.throws(() => {
                fs.statSync(prefix("output/factor-bundle/noempty/b.css"));
            });
                    
            expect(read("./factor-bundle/noempty/noempty.css")).toMatchSnapshot();
            expect(read("./factor-bundle/noempty/a.css")).toMatchSnapshot();
        });

        it("should output empty css files when asked", async () => {
            const build = browserify([
                    prefix("specimens/factor-bundle/empty/a.js"),
                    prefix("specimens/factor-bundle/empty/b.js"),
                ]);
            
            build.plugin(plugin, {
                css   : prefix("output/factor-bundle/empty/empty.css"),
                empty : true,
            });

            build.plugin(require("factor-bundle"), {
                outputs : [
                    prefix("output/factor-bundle/empty-a.js"),
                    prefix("output/factor-bundle/empty-b.js"),
                ],
            });
            
            await bundle(build);

            expect(read("./factor-bundle/empty/empty.css")).toMatchSnapshot();
            expect(read("./factor-bundle/empty/a.css")).toMatchSnapshot();
        });
    });
});
