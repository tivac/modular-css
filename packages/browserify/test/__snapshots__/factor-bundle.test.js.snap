// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/browserify.js factor-bundle should avoid outputting empty css files by default 1`] = `
"/* packages/browserify/test/specimens/folder/folder.css */
.mcd19f1319_folder { margin: 2px; }
"
`;

exports[`/browserify.js factor-bundle should avoid outputting empty css files by default 2`] = `
"/* packages/browserify/test/specimens/local.css */
.mcc4e2a8e1_booga { background: green; }
/* packages/browserify/test/specimens/start.css */
.mc1c78395f_booga { color: red; background: blue; }
.mc1c78395f_tooga { border: 1px solid white; }
"
`;

exports[`/browserify.js factor-bundle should be supported 1`] = `
"require=(function(){function r(e,n,t){function o(i,f){if(!n[i]){if(!e[i]){var c=\\"function\\"==typeof require&&require;if(!f&&c)return c(i,!0);if(u)return u(i,!0);var a=new Error(\\"Cannot find module '\\"+i+\\"'\\");throw a.code=\\"MODULE_NOT_FOUND\\",a}var p=n[i]={exports:{}};e[i][0].call(p.exports,function(r){var n=e[i][1][r];return o(n||r)},p,p.exports,r,e,n,t)}return n[i].exports}for(var u=\\"function\\"==typeof require&&require,i=0;i<t.length;i++)o(t[i]);return o}return r})()({5:[function(require,module,exports){
module.exports = {
    \\"folder\\": \\"white\\",
    \\"one\\": \\"red\\",
    \\"two\\": \\"blue\\",
    \\"booga\\": \\"mcc4e2a8e1_booga\\",
    \\"looga\\": \\"mcc4e2a8e1_booga mcc4e2a8e1_looga\\"
};
},{\\"./packages/browserify/test/specimens/folder/folder.css\\":4}],3:[function(require,module,exports){
require(\\"../../folder/folder.css\\");

},{\\"../../folder/folder.css\\":4}],4:[function(require,module,exports){
module.exports = {
    \\"folder\\": \\"mcd19f1319_folder\\"
};
},{}]},{},[]);
"
`;

exports[`/browserify.js factor-bundle should be supported 2`] = `
"/* packages/browserify/test/specimens/folder/folder.css */
.mcd19f1319_folder { margin: 2px; }
/* packages/browserify/test/specimens/local.css */
.mcc4e2a8e1_booga { background: green; }
"
`;

exports[`/browserify.js factor-bundle should be supported 3`] = `
"/* packages/browserify/test/specimens/start.css */
.mc1c78395f_booga { color: red; background: blue; }
.mc1c78395f_tooga { border: 1px solid white; }
"
`;

exports[`/browserify.js factor-bundle should output empty css files when asked 1`] = `
"/* packages/browserify/test/specimens/folder/folder.css */
.mcd19f1319_folder { margin: 2px; }
"
`;

exports[`/browserify.js factor-bundle should output empty css files when asked 2`] = `
"/* packages/browserify/test/specimens/local.css */
.mcc4e2a8e1_booga { background: green; }
/* packages/browserify/test/specimens/start.css */
.mc1c78395f_booga { color: red; background: blue; }
.mc1c78395f_tooga { border: 1px solid white; }
"
`;

exports[`/browserify.js factor-bundle should properly handle files w/o dependencies 1`] = `
"/* packages/browserify/test/specimens/folder/folder.css */
.mcd19f1319_folder { margin: 2px; }
/* packages/browserify/test/specimens/local.css */
.mcc4e2a8e1_booga { background: green; }
"
`;

exports[`/browserify.js factor-bundle should properly handle files w/o dependencies 2`] = `
"/* packages/browserify/test/specimens/simple.css */
.mc459e927d_wooga { color: red; }
"
`;

exports[`/browserify.js factor-bundle should support files w/o commonalities 1`] = `
"/* packages/browserify/test/specimens/simple.css */
.mc459e927d_wooga { color: red; }
"
`;

exports[`/browserify.js factor-bundle should support files w/o commonalities 2`] = `
"/* packages/browserify/test/specimens/blue.css */
.mcd53407d1_wooga { color: blue; }
"
`;

exports[`/browserify.js factor-bundle should support relative paths within factor-bundle files 1`] = `
"/* packages/browserify/test/specimens/folder/folder.css */
.mcd19f1319_folder { margin: 2px; }
/* packages/browserify/test/specimens/local.css */
.mcc4e2a8e1_booga { background: green; }
"
`;

exports[`/browserify.js factor-bundle should support relative paths within factor-bundle files 2`] = `
"/* packages/browserify/test/specimens/relative.css */
.mcdf61ec72_wooga {
    color: red;
    background: url(\\"../../../specimens/folder/to.png\\");
}
/* packages/browserify/test/specimens/start.css */
.mc1c78395f_booga { color: red; background: blue; }
.mc1c78395f_tooga { border: 1px solid white; }
"
`;
